shapely/constructive.py

```diff
@@ -887,10 +887,23 @@ def simplify(geometry, tolerance, preserve_topology=True, **kwargs):
 def snap(geometry, reference, tolerance, **kwargs):
     """Snaps an input geometry to reference geometry's vertices.
 
-    The tolerance is used to control where snapping is performed.
+    Vertices of the first geometry are snapped to vertices of the second.
+    geometry. The input geometries are not modified.
     The result geometry is the input geometry with the vertices snapped.
     If no snapping occurs then the input geometry is returned unchanged.
 
+    The tolerance is used to control where snapping is performed.
+    A snap distance tolerance is used to control where snapping is performed.
+
+    Snapping one geometry to another can improve robustness for overlay
+    operations by eliminating nearly-coincident edges
+    (which cause problems during noding and intersection calculation).
+
+    Too much snapping can result in invalid topology being created,
+    so the number and location of snapped vertices is decided using
+    heuristics to determine when it is safe to snap.
+    This can result in some potential snaps being omitted, however.
```

Comment:
Does this mean that snapping will not always happen, even when within the tolerance distance, to avoid invalid geometries? That's something the underlying algorithm does automatically?

Reply:
This is copy pasted from geos docs:
https://libgeos.org/doxygen/classgeos_1_1operation_1_1overlay_1_1snap_1_1GeometrySnapper.html#details
I'm not sure I fully understand the logic of the tolerance. I couldn't find a better documentation of what exactly it means.

Reply:
This docstring originally came from [JTS](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/operation/overlay/snap/GeometrySnapper.html) 

My understanding is that the snapping algorithm will avoid using potential snap targets if they would lead to an invalid geometry.

Reply:
@brendan-ward JTS is BSD-style licensed, so if this text also is present there, I don't think we have to worry too much about licensing